application.go

```diff
@@ -57,3 +65,127 @@ func (ac *ApplicationClient) GetApplication(
 
 	return res.Application, nil
 }
+
+// GetApplicationsDelegatingToGateway returns the application addresses that are
+// delegating to the given gateway address.
+func (ac *ApplicationClient) GetApplicationsDelegatingToGateway(
+	ctx context.Context,
+	gatewayAddress string,
+	queryHeight uint64,
+) ([]string, error) {
+	allApplications, err := ac.GetAllApplications(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("GetApplicationsDelegatingToGateway: error getting all applications: %w", err)
+	}
+
+	gatewayDelegatingApplications := make([]string, 0)
+	for _, application := range allApplications {
+		appRing := ApplicationRing{Application: application, SessionEndBlock: queryHeight}
+		// Get the gateways that are delegated to the application
+		// at the query height and check if the given gateway address is in the list.
+		gatewaysDelegatedTo := appRing.ringAddressesAtBlock()
+		if slices.Contains(gatewaysDelegatedTo, gatewayAddress) {
+			// The application is delegating to the given gateway address, add it to the list.
+			gatewayDelegatingApplications = append(gatewayDelegatingApplications, application.Address)
+		}
+	}
+
+	return gatewayDelegatingApplications, nil
+}
+
+type ApplicationRing struct {
+	types.Application
+	PublicKeyFetcher
+	SessionEndBlock uint64
+}
+
+// GetRing returns the ring for the application.
+// The ring is created using the application's public key and the public keys of
+// the gateways that are currently delegated from the application.
+func (a ApplicationRing) GetRing(
+	ctx context.Context,
+) (addressRing *ring.Ring, err error) {
+	if a.PublicKeyFetcher == nil {
+		return nil, errors.New("GetRing: Public Key Fetcher not set")
+	}
+
+	if a.SessionEndBlock <= 0 {
+		return nil, errors.New("GetRing: Current Height not set")
+	}
+
+	// Get the gateway addresses that are delegated from the application
+	// at the query height.
+	currentGatewayAddresses := a.ringAddressesAtBlock()
+
+	ringAddresses := make([]string, 0)
```

Comment:
Could you please add a `TODO` to use a stateless ring constructor from `poktroll` once available?

Reply:
Thank you for the review. fixed.